MIDAS/src/hardware/Pyro.cpp

```diff
@@ -0,0 +1,114 @@
+#include "sensors.h"
+#include "pins.h"
+
+#include "TCAL9539-Q1.h"
+
+#define MAXIMUM_TILT_ANGLE 20
+
+GPIOTCAL9539 gpio_expander{};
+
+/**
+ * "Initializes" the pyro thread. The main initialization will be done by the GPIO expander, so the pyro thread doesn't
+ * have to do anything special.
+*/
+ErrorCode PyroThread::init() {
+    bool gpio_init = gpio_expander.init();
+    if(gpio_init) {
+        return ErrorCode::NoError;
+    } else {
+        return ErrorCode::PyroGPIOCouldNotBeInitialized;
+    }
+    
+}
+
+/**
+ * Upper stage only!
+ * Returns a new pyro struct, with data depending on whether or not each pyro channel should be firing.
+ * Fires channels by setting their pin on the GPIO.
+*/
+Pyro PyroThread::tick_upper(FSMState fsm_state, Orientation orientation) {
+    Pyro new_pyro = Pyro();
+
+    // If the state is IDLE or any state after that, we arm the global arm pin
+    if(fsm_state.curr_state >= FSM_state::STATE_IDLE) {
+        new_pyro.global_armed = true;
+        gpio_expander.write(PYRO_GLOBAL_ARM_PIN, HIGH);
+    }
+
+    // Fire "Pyro A" to ignite sustainer
+    // Additionally, check if orientation allows for firing
+    if(fsm_state.curr_state == FSM_state::STATE_SUSTAINER_IGNITION && std::abs(orientation.pitch) < MAXIMUM_TILT_ANGLE) {
```

Comment:
we would also have to check roll as well i believe. The axes might be fucked up here and we'd have to change after we do some testing because it depends on how the bno is oriented, but you would have to check 2 axes for sure